File: 1559-detect-cycles-in-2d-grid/solution.py
class UnionFind:
    def __init__(self, n):
        self.parent = list(range(n))
        self.rank = [0] * n

    def find(self, i):
        if self.parent[i] == i:
            return i
        self.parent[i] = self.find(self.parent[i])
        return self.parent[i]

    def union(self, i, j):
        root_i = self.find(i)
        root_j = self.find(j)
        if root_i != root_j:
            if self.rank[root_i] < self.rank[root_j]:
                self.parent[root_i] = root_j
            elif self.rank[root_i] > self.rank[root_j]:
                self.parent[root_j] = root_i
            else:
                self.parent[root_i] = root_j
                self.rank[root_j] += 1
            return True
        return False

class Solution:
    def containsCycle(self, grid: list[list[str]]) -> bool:
        rows = len(grid)
        cols = len(grid[0])
        uf = UnionFind(rows * cols)

        for r in range(rows):
            for c in range(cols):
                # We only need to check right and down neighbors to cover all edges
                # without processing the same edge twice.
                
                # Check Right Neighbor
                if c + 1 < cols and grid[r][c] == grid[r][c + 1]:
                    u = r * cols + c
                    v = r * cols + (c + 1)
                    if uf.find(u) == uf.find(v):
                        return True
                    uf.union(u, v)
                
                # Check Down Neighbor
                if r + 1 < rows and grid[r][c] == grid[r + 1][c]:
                    u = r * cols + c
                    v = (r + 1) * cols + c
                    if uf.find(u) == uf.find(v):
                        return True
                    uf.union(u, v)
        
        return False

File: 1559-detect-cycles-in-2d-grid/solution.cpp
#include <vector>

using namespace std;

class Solution {
public:
    bool containsCycle(vector<vector<char>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        vector<vector<bool>> visited(m, vector<bool>(n, false));
        
        int dr[] = {0, 0, 1, -1};
        int dc[] = {1, -1, 0, 0};
        
        // Helper function for DFS
        auto dfs = [&](auto self, int r, int c, int pr, int pc, char target) -> bool {
            visited[r][c] = true;
            
            for (int i = 0; i < 4; ++i) {
                int nr = r + dr[i];
                int nc = c + dc[i];
                
                // Boundary and character match check
                if (nr >= 0 && nr < m && nc >= 0 && nc < n && grid[nr][nc] == target) {
                    // If neighbor is visited and is NOT the parent, cycle detected
                    if (visited[nr][nc]) {
                        if (nr != pr || nc != pc) return true;
                    } else {
                        if (self(self, nr, nc, r, c, target)) return true;
                    }
                }
            }
            return false;
        };
        
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (!visited[i][j]) {
                    // Start DFS for each unvisited component
                    if (dfs(dfs, i, j, -1, -1, grid[i][j])) return true;
                }
            }
        }
        
        return false;
    }
};